### PR TITLE
Implement CollectionTracker for React Native

### PR DESCRIPTION
**Pull request by rockapot**
# CollectionTracker

This implementation of `CollectionTracker` is designed to be a centralized configuration mechanism for list views that are integrated with Promoted. Integration is done like this for functional components:
```ts
const MyListComponent: FunctionComponent<MyListProps> = (props) => {
   // Wrap the list component before use.
   const TrackedList = useCollectionTracker({
     contentCreator: (item) => ({
       contentId: item.internalId.toString(),
       insertionId: item.insertionId,
       name: item.displayName,
     }),
     sourceType: ImpressionSourceType.Delivery,
   })(FlatList)
   // Instantiate the tracked list.
   return (
     <TrackedList
       data={...}
       renderItem={...}
       {...rest}
     />
   )
 }
```
A similar pattern can be used for class components:
```ts
// This defines a tracked list component for use later
const TrackedList = CollectionTracker({
  contentCreator: (item) => ({
    contentId: item.id?.toString(),
    insertionId: item.insertionId,
    name: item.name,
  }),
})(FlatList)

class MyContentList extends PureComponent<MyContentListProps> {
  public render() {
    return (
      <TrackedList // Use the tracked list component here
        data={contents}
        renderItem={this.renderItem}
        {...props}
      />
    )
  }
}
```

# Supplemental Changes

In addition to the implementation of `CollectionTracker`, made the following changes, some of which mirror the changes in https://github.com/promotedai/ios-metrics-sdk/pull/151.

1. Reduce the number of `log*Action` methods in the interface. Instead, use the `ActionType` enum to specify the different types, and accept a content argument for every method.
2. Add support for `collectionViewActionDidOccur` in the native module. This tracks an action in an associated collection view. It's here that the benefit of `CollectionTracker` occurs. We can verify the insertion IDs against whether we're expecting them or not (based on `ImpressionSourceType`), ensure that impression IDs are set correctly on action events, and eventually support more advanced features.
3. As a stylistic nit, removed semicolons from TypeScript code.

# Caveats

This PR does not yet support the following features, which we plan to add soon:

1. Enhanced view ID logging. The view IDs here still being generated using previous means. Eventually, we'll ensure that every event logged from the same CollectionTracker has the same view ID.
2. Support for additional actions on list items that aren't `Navigate`. For example, tapping on a component inside the list item to Like/Unlike content. These use cases can be more complex because the actual action may not occur immediately, so we'll have to figure out how to pass CollectionTracker state around, in a way that's not too burdensome for clients.

### CODE REVIEW

**Comment by rockapot** [comment]
Note that checks for this PR won't pass until https://github.com/promotedai/ios-metrics-sdk/pull/151 is merged.

**Review by andrewdyates — APPROVED** [review]

File: src/PromotedMetricsType.tsx [hunk 1]
```diff
@@ -18,67 +29,85 @@ export type PromotedMetricsType = {
...
    */
-  startSessionAndLogSignedOutUser(): void;
+  startSessionAndLogSignedOutUser(): void
 
   // Impression logging
 
   /**
    * Logs an impression for given content.
-   * Typically, you would call useImpressionTracker() for use with
+   * Typically, you would call withCollectionTracker() for use with
```
Comment (danbosnichill) [inline src/PromotedMetricsType.tsx:45]: Interesting change.  Will clients not useImpressionTracker at all anymore?

File: src/PromotedMetricsType.tsx [hunk 2]
```diff
@@ -18,67 +29,85 @@ export type PromotedMetricsType = {
...
+   */
+  logNavigateAction(content: Object): void
+
+  /**
+   * Logs a clickthrough for details about given content.
+   *
+   * @param content content whose details are requested
+   * @param screenName name of screen that will display content details
+   */
+   logNavigateActionWithScreenName(content: Object, screenName: string): void
```
Comment (danbosnichill) [inline src/PromotedMetricsType.tsx:74]: nit indentation

**Review by danbosnichill — APPROVED** [review]

File: src/PromotedMetricsType.tsx (hunk 1 above)
Comment (rockapot) [reply src/PromotedMetricsType.tsx:45]: ImpressionTracker will still be around for cases where we need to use only impression tracking, but will not be the default/recommended choice.

File: src/CollectionTracker.tsx [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+ * viewability tracking properties to the list, then they are merged into
+ * the `viewabilityConfigCallbackPairs` property. The aforementioned
+ * property will include both viewability tracking for Promoted impressions
+ * and your existing viewability tracking.
+ *
+ * @returns Wrapped component to use as list component
+ */
+export function CollectionTracker<P extends CollectionTrackerProps>({
+  contentCreator,
+  sourceType = ImpressionSourceType.ClientBackend
```
[the remark this answers is not in the document]
Comment (rockapot) [reply src/CollectionTracker.tsx]: Done.

File: src/PromotedMetricsType.tsx (hunk 2 above)
Comment (rockapot) [reply src/PromotedMetricsType.tsx]: Done.

**Pull request merged**